ftdc/custom_format.go

```diff
@@ -0,0 +1,549 @@
+			prevValues = nil
+
+			//
+			continue
+		} else if schema == nil {
+			return nil, errors.New("First byte of FTDC data must be the magic one")
+		}
+
+		// This FTDC document is a metric document. Read the "diff bits" that describe which metrics
+		// have changed since the prior metric document. Note, the reader is positioned on the
+		// "packed byte" where the first bit is not a diff bit. `readDiffBits` must account for
+		// that.
+		diffedFields := readDiffBits(reader, schema)
+
+		// The next eight bytes after the diff bits is the time in nanoseconds since the 1970 epoch.
+		var dataTime int64
+		binary.Read(reader, binary.BigEndian, &dataTime)
+
+		// Read the payload. There will be one float32 value for each diff bit set to `1`, i.e:
+		// `len(diffedFields)`.
+		data, err := readData(reader, schema, diffedFields, prevValues)
+		if err != nil {
+			return ret, err
+		}
+
+		// The old `prevValues` is no longer needed. Set the `prevValues` to the new hydrated
+		// `data`.
+		prevValues = data
+
+		// Construct a `Datum` that hydrates/merged the full set of float32 metrics with the metric
+		// names as written in the most recent schema document.
+		ret = append(ret, Datum{
+			Time: dataTime,
+			Data: schema.Hydrate(data),
+		})
+	}
+
+	return ret, nil
+}
+
+// readSchema expects to be positioned on the beginning of a json list data type (a left square
+// bracket `[`) and consumes bytes until that list (of strings) is complete.
+//
+// readSchema returns the described schema and a new reader that's positioned on the first byte of
+// the next ftdc document.
+func readSchema(reader *bufio.Reader) (*Schema, *bufio.Reader) {
+	decoder := json.NewDecoder(reader)
+	if !decoder.More() {
+		panic("no json")
+	}
+
+	// While the FTDC metrics persisted has structure, we flatten the metric names into a single
+	// list of strings. We use dots (`.`) to signify nesting. Metric names with dots will result in
+	// an ambiguous parsing.
+	var fields []string
+	if err := decoder.Decode(&fields); err != nil {
+		panic(err)
+	}
+
+	// The JSON decoder can consume bytes from the input `reader` that are beyond the end of the
+	// json data. Those unused bytes are accessible via the `decoder.Buffered` call. Assemble a new
+	// reader with the remaining bytes from the decoder, followed by the remaining bytes from the
+	// input `reader`.
+	retReader := bufio.NewReader(io.MultiReader(decoder.Buffered(), reader))
+
+	// Consume a newline character. The JSON Encoder will unconditionally append a newline that the
+	// JSON decoder will not* consume. This is a sharp edge of the Golang JSON API.
+	ch, _ := retReader.ReadByte()
+	if ch != '\n' {
+		panic("not a newline")
+	}
+
+	// We now have fields, e.g: ["metric1.Foo", "metric1.Bar", "metric2.Foo"]. The `mapOrder` should
+	// be ["metric1", "metric2"]. It's undefined behavior for a `mapOrder` metric name key to be
+	// split around a different metric name. E.g: ["metric1.Alpha", "metric2.Beta",
```

Comment:
I don't think I would have parsed that the schema columns MUST have one level of nesting. Can we write that somewhere in the explanation? I would have thought that writing ["alpha", "matric1.beta", "gamma"] would be acceptable.

Reply:
Similar where I'd like to talk offline about options. The file format doesn't require this at all. But the usage (being `map[string]any`) does result in this small translation. Maybe I shouldn't have put it in this "custom format" code and rather left it as part of the caller's responsibility.

Reply:
cool, lets chat about this in the in person chat